### ISSUE TITLE
Console/Class lib project templates use latest C# language idioms

### ISSUE THREAD

**Issue by DamianEdwards** [dotnet/templating#3360]
In .NET 6, the included C# project templates will use the latest language idioms.

See the [parent Epic](https://github.com/dotnet/templating/issues/3359) for details.

Console template: 
- [x] remove app template - https://github.com/dotnet/templating/pull/3424
- [x] top-level program - https://github.com/dotnet/templating/pull/3424
- [x] global usings - https://github.com/dotnet/templating/pull/3424
- [x] nullable - https://github.com/dotnet/templating/pull/3424

Classlib template:
- [x] global usings - https://github.com/dotnet/templating/pull/3424
- [x] nullable https://github.com/dotnet/templating/pull/3173 https://github.com/dotnet/templating/pull/3299 https://github.com/dotnet/templating/pull/3396
- [x] file-scoped namespaces- https://github.com/dotnet/templating/pull/3395

Issues:
- [x] https://github.com/dotnet/templating/issues/3449

@KathleenDollard

**Comment by vlada-shubina** [comment]
@DamianEdwards  

Couple things here: 

1. Recently we've created new app template (https://github.com/dotnet/templating/issues/3248). This template suppose to replace console template. It already has most of the features in and we are adding global usings in https://github.com/dotnet/templating/issues/3344. I can add `async Main` in same time.

2. Nullables were already added to classlib template: https://github.com/dotnet/templating/issues/2979. I can add file scoped templates and global usings if needed. 

3. Just double checking - any changes needed for other languages (F#, VB)?

**Comment by KathleenDollard** [comment]
@DamianEdwards There is an enormous amount of tutorials, instructions, books, etc in the wild that has been created over the years. This includes everything from high school curriculum to SO posts and is used by the people that are introducing .NET for us. Given the amount of this information and the fact it fundamentally not working affects our newest customers as they are just trying .NET, we decided that a new template was least disruptive. In addition, the name "console" has meaning only from the perspective of the tech, not naïve intuition. We decided to "freeze" Console, and thus did not add nullable.

For these reasons we added a new "app" template with the expectation that all beginner instructions etc. will change to this template over (hopefully short) time allowing us to make this change without disrupting our new users and have a better name. In addition to the scenario where a teacher hands out instructions that make no sense, someone could rebuild their laptop a week before finals to do some end semester work and be unable to do it. We can instead teach teachers about using these new features and allow them to introduce namespaces and class at the time they think appropriate. Of course we update things that can be disruptive, but this massive change to Console would disrupt the wrong people at the wrong time. 

Something else has taken a great deal of time over the last several days, and thus the blog post for templates in Preview 6 was delayed.

**Comment by KathleenDollard** [comment]
@vlada-shubina There is another change we are working on that will affect global usings overall. So hold off on that change to the app template please.

**Comment by DamianEdwards** [comment]
@KathleenDollard thanks for the details. Given that context, are we planning on removing console from the default list shown when `dotnet new` is invoked without arguments? The request here is that all our templates are updated to reflect the latest and greatest language idioms while acknowledging that this could cause friction in some cases.  Given the specific concerns RE console and education however, perhaps a compromise of not showing it in the default set is acceptable.

@vlada-shubina no need to add `async Main` unless the template is already calling `Task`-returning methods, or calling methods that have `Task`-returning equivalents.

**Comment by vlada-shubina** [comment]
@DamianEdwards @KathleenDollard File-scoped namespaces were added in https://github.com/dotnet/templating/pull/3395, however these templates cannot be built at the moment, therefore we cannot merge it until it is supported for build.

**Comment by DamianEdwards** [comment]
@vlada-shubina thanks, [this is the compiler issue to watch](https://github.com/dotnet/roslyn/issues/49000) for them landing the file-scoped namespaces feature in preview.7

**Comment by vlada-shubina** [comment]
@DamianEdwards on this thread: we are removing app template https://github.com/dotnet/templating/issues/3423 now and porting all features of it to console net6.0 template.

@DamianEdwards @KathleenDollard what should be behavior or the features with regards to language version: we have a parameter `langVersion` that sets C# language version to use. 
- global usings are not available before 10.0
- nullable are not available before 8.0

Should we make a condition on those or should the template just fail in case lang version is different from 10.0?

**Comment by vlada-shubina** [comment]
@DamianEdwards @KathleenDollard another issue blocking PRs with global usings: global usings are not working for language version other than 10.0. The project on language version below 10.0 cannot be compiled. 
This blocks number of tests.

Example:
```
> dotnet new console --langVersion 9.0
(succeeds)

> dotnet build
    Microsoft (R) Build Engine version 17.0.0-preview-21359-04+d150e93ff for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\obj\Debug\net6.0\258e9877-0085-44ec-91bb-419c11a5f7a3.ImplicitNamespaceImports.cs(2,1): error CS8773: Feature 'global using directive' is not available in C# 9.0. Please use language version 10.0 or greater. [C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\258e9877-0085-44ec-91bb-419c11a5f7a3.csproj]
C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\obj\Debug\net6.0\258e9877-0085-44ec-91bb-419c11a5f7a3.ImplicitNamespaceImports.cs(3,1): error CS8773: Feature 'global using directive' is not available in C# 9.0. Please use language version 10.0 or greater. [C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\258e9877-0085-44ec-91bb-419c11a5f7a3.csproj]
C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\obj\Debug\net6.0\258e9877-0085-44ec-91bb-419c11a5f7a3.ImplicitNamespaceImports.cs(4,1): error CS8773: Feature 'global using directive' is not available in C# 9.0. Please use language version 10.0 or greater. [C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\258e9877-0085-44ec-91bb-419c11a5f7a3.csproj]
C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\obj\Debug\net6.0\258e9877-0085-44ec-91bb-419c11a5f7a3.ImplicitNamespaceImports.cs(5,1): error CS8773: Feature 'global using directive' is not available in C# 9.0. Please use language version 10.0 or greater. [C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\258e9877-0085-44ec-91bb-419c11a5f7a3.csproj]
C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\obj\Debug\net6.0\258e9877-0085-44ec-91bb-419c11a5f7a3.ImplicitNamespaceImports.cs(6,1): error CS8773: Feature 'global using directive' is not available in C# 9.0. Please use language version 10.0 or greater. [C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\258e9877-0085-44ec-91bb-419c11a5f7a3.csproj]
C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\obj\Debug\net6.0\258e9877-0085-44ec-91bb-419c11a5f7a3.ImplicitNamespaceImports.cs(7,1): error CS8773: Feature 'global using directive' is not available in C# 9.0. Please use language version 10.0 or greater. [C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\258e9877-0085-44ec-91bb-419c11a5f7a3.csproj]
C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\obj\Debug\net6.0\258e9877-0085-44ec-91bb-419c11a5f7a3.ImplicitNamespaceImports.cs(8,1): error CS8773: Feature 'global using directive' is not available in C# 9.0. Please use language version 10.0 or greater. [C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\258e9877-0085-44ec-91bb-419c11a5f7a3.csproj]

Build FAILED.

C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\obj\Debug\net6.0\258e9877-0085-44ec-91bb-419c11a5f7a3.ImplicitNamespaceImports.cs(2,1): error CS8773: Feature 'global using directive' is not available in C# 9.0. Please use language version 10.0 or greater. [C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\258e9877-0085-44ec-91bb-419c11a5f7a3.csproj]
C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\obj\Debug\net6.0\258e9877-0085-44ec-91bb-419c11a5f7a3.ImplicitNamespaceImports.cs(3,1): error CS8773: Feature 'global using directive' is not available in C# 9.0. Please use language version 10.0 or greater. [C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\258e9877-0085-44ec-91bb-419c11a5f7a3.csproj]
C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\obj\Debug\net6.0\258e9877-0085-44ec-91bb-419c11a5f7a3.ImplicitNamespaceImports.cs(4,1): error CS8773: Feature 'global using directive' is not available in C# 9.0. Please use language version 10.0 or greater. [C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\258e9877-0085-44ec-91bb-419c11a5f7a3.csproj]
C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\obj\Debug\net6.0\258e9877-0085-44ec-91bb-419c11a5f7a3.ImplicitNamespaceImports.cs(5,1): error CS8773: Feature 'global using directive' is not available in C# 9.0. Please use language version 10.0 or greater. [C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\258e9877-0085-44ec-91bb-419c11a5f7a3.csproj]
C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\obj\Debug\net6.0\258e9877-0085-44ec-91bb-419c11a5f7a3.ImplicitNamespaceImports.cs(6,1): error CS8773: Feature 'global using directive' is not available in C# 9.0. Please use language version 10.0 or greater. [C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\258e9877-0085-44ec-91bb-419c11a5f7a3.csproj]
C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\obj\Debug\net6.0\258e9877-0085-44ec-91bb-419c11a5f7a3.ImplicitNamespaceImports.cs(7,1): error CS8773: Feature 'global using directive' is not available in C# 9.0. Please use language version 10.0 or greater. [C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\258e9877-0085-44ec-91bb-419c11a5f7a3.csproj]
C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\obj\Debug\net6.0\258e9877-0085-44ec-91bb-419c11a5f7a3.ImplicitNamespaceImports.cs(8,1): error CS8773: Feature 'global using directive' is not available in C# 9.0. Please use language version 10.0 or greater. [C:\Users\vshubina\AppData\Local\Temp\TemplateEngine.Tests\258e9877-0085-44ec-91bb-419c11a5f7a3\258e9877-0085-44ec-91bb-419c11a5f7a3.csproj]
    0 Warning(s)
    7 Error(s)

Time Elapsed 00:00:01.96

**Comment by DamianEdwards** [comment]
@vlada-shubina yes global usings requires C# 10. @JunTaoLuo do the implicit namespace conditions in the SDK not include a check of LangVersion if it's defined?

**Comment by JunTaoLuo** [comment]
I'm hesitant about a LangVersion check. Specifically, I couldn't find a good way of expressing the logic "LangVersion >= 10.0". 

LangVersion can be set to non-numerical values such as `latest`, `latestMajor`, `default`, `preview` each of which may resolve as different numerical versions. I think all this logic is encapsulated in the compiler at the moment and if the resolution can be made available in an MSBuild task, I wouldn't mind a LangVersion check in the SDK. As it stands now, I'm not sure it's a great idea to replicate the compiler logic for resolving the raw LangVersion since it may drift from the compiler and hard to ensure parity.

As for the tests being blocked, could we disable the feature with `DisableImplicitNamespaceImports` before building. That is what we envisioned for projects that have the new TFM (>net6.0) but specifically sets the `LangVersion` to a lower level. I'm not sure how we can do that in the template though since it would also require us to understand `LangVersion` semantically.

Another idea is that we should output a better error message somehow that indicates if CS8773 occurs, `DisableImplicitNamespaceImports` can be set to disable this feature.

**Comment by vlada-shubina** [comment]
@JunTaoLuo I see, we have the similar problem here. We have a `--langVersion` option, which allows setting the value of `<LangVersion>` property in csproj file when running `dotnet new`. We also don't have a good way to manage conditions on that for the new features.
The draft was here: https://github.com/dotnet/templating/pull/3424/commits/4a16adff0e19a4d841ea7b028d151dcb06689a3 however it doesn't take into account `latest, preview` etc values.  

@DamianEdwards have you discussed this issue before? dotnet/templating is not the only repo using `langVersion` property, at least wpf and winforms are using it as well. It would be good to align on strategy so we do same thing across the repos.

**Comment by DamianEdwards** [comment]
Yeah seems we'll have to condition this in the templates that have a `--langVersion` option (the web templates do not).

Your enumeration seems like the correct approach.

**Comment by vlada-shubina** [comment]
@DamianEdwards all tasks in this issue have been completed, closing. Feel free to reopen if needed.

**Comment by jcouv** [comment]
The file-scoped namespaces portion of this change is causing some pain, as the IDE experience is not yet fully baked for this feature. Added you to the email thread to discuss temporarily reverting that change. Thanks
